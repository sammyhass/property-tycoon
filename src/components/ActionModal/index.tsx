import { useGameContext } from '@/hooks/useGameContext';
import { formatPrice } from '@/util/formatPrice';
import {
  Box,
  Button,
  Divider,
  Flex,
  Heading,
  Modal,
  ModalBody,
  ModalCloseButton,
  ModalContent,
  ModalHeader,
  ModalOverlay,
  Stack,
  Text,
} from '@chakra-ui/react';
import { CardAction, CardType } from '@prisma/client';
import React, { useState } from 'react';
import BoardSpace from '../Board/spaces';

// Different types of action modal we can use for our modal.
export type ActionType =
  | 'BUY' // Buying a property
  | 'MORTGAGE'
  | 'PAY_RENT' // Pay rent
  | 'ROLL' // Roll dice
  // Taking Cards
  | 'TAKE_POT_LUCK'
  | 'TAKE_OPPORTUNITY_KNOCKS'
  // ---
  | 'GO_TO_JAIL' // Player is sent to jail
  | 'FREE_PARKING' // Land on a Free Parking space
  | 'GO'; // Land on a Go space

interface ActionModalProps {
  action: ActionType | null;
  onClose: () => void;
  isOpen: boolean;
}

const actionModalComponents: Record<ActionType, React.FC<ActionModalProps>> = {
  BUY: () => <ActionModal.Buy />,
  MORTGAGE: () => <div>Mortgage</div>,
  PAY_RENT: () => <div>Pay Rent</div>,
  ROLL: props => <ActionModal.Roll {...props} />,
  GO_TO_JAIL: () => <div>Jail</div>,
  TAKE_POT_LUCK: () => <ActionModal.TakeCard cardType={'POT_LUCK'} />,
  TAKE_OPPORTUNITY_KNOCKS: () => (
    <ActionModal.TakeCard cardType="OPPORTUNITY_KNOCKS" />
  ),
  FREE_PARKING: () => <div>Free Parking</div>,
  GO: () => <div>GO!</div>,
};

export default function ActionModal(props: ActionModalProps) {
  const ActionComponent = props.action
    ? (actionModalComponents[props.action] as React.FC<ActionModalProps>)
    : null;

  return (
    <Modal onClose={props.onClose} isOpen={props.isOpen}>
      <ModalOverlay />
      <ModalContent>
        <ModalHeader>{props.action}</ModalHeader>
        <ModalCloseButton />
        <ModalBody>
          {ActionComponent && <ActionComponent {...props} />}
        </ModalBody>
      </ModalContent>
    </Modal>
  );
}

// Content for our action modal when a player is rolling the dice.
ActionModal.Roll = ({ onClose }: ActionModalProps) => {
  const { currentPlayer, move } = useGameContext();

  const [isRolling, setIsRolling] = useState(false);
  const [roll, setRoll] = useState(0);

  // Roll the dice.
  const handleRoll = () => {
    setIsRolling(true);

    if (!currentPlayer) return;
    // Choose a random number between 1 and 6
    const roll = Math.floor(Math.random() * 6) + 1;

    setTimeout(() => {
      move(currentPlayer?.token, roll);
      setRoll(roll);
      setIsRolling(false);
    }, 1000);

    if (!currentPlayer) {
      return <></>;
    }
  };
  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="4xl">ðŸŽ²</Heading>
      {isRolling && <Heading size="md">Rolling...</Heading>}
      {roll > 0 && <Heading size="sm">Dice landed on {roll}</Heading>}
      <Button onClick={() => handleRoll()}>Roll</Button>
    </Flex>
  );
};

// Content of our action modal for taking a card (pot luck or opportunity knocks)
ActionModal.TakeCard = ({ cardType }: { cardType: CardType }) => {
  const { takeCard } = useGameContext();

  const [cardAction, setCardAction] = useState<CardAction | null>(null);

  const handleTakeCard = () => {
    const action = takeCard(cardType);
    if (action) {
      setCardAction(action);
    }
  };

  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="md">
        Pick a {cardType === 'POT_LUCK' ? 'Pot Luck' : 'Opportunity Knocks'}{' '}
        Card
      </Heading>
      <pre>{JSON.stringify(cardAction, null, 2)}</pre>
      <Button onClick={() => handleTakeCard()}>Take</Button>
    </Flex>
  );
};

ActionModal.Buy = () => {
  const { gameSettings, currentPlayer, state, buy, hideActionModal } =
    useGameContext();
  const [buying, setBuying] = useState(false);

  // Logic: the only way to buy a property is if it is your turn and you have landed on it.

  if (!currentPlayer) return <></>;

  const space = gameSettings?.BoardSpaces.find(
    space => space.board_position === state[currentPlayer?.token]?.pos
  );

  const property = gameSettings?.Properties.find(
    property => property.id === space?.property_id
  );

  if (!property) return <></>;

  const handleBuy = () => {
    setBuying(true);
    buy(currentPlayer?.token, property.id);
    setTimeout(() => {
      hideActionModal();
      setBuying(false);
    }, 1000);
  };

  return (
    <Box>
      <Flex direction={'column'} justify={'center'} align="center">
        <BoardSpace.Property property={property} />
      </Flex>
      <Divider my="15px" />
      <Box my="10px" p="10px" bg="#eee" borderRadius={'8px'}>
        <Heading size="md">
          {property.name} - {formatPrice(property.price ?? 0)}
        </Heading>
        <Stack>
          <Text p="0" fontWeight={'600'}>
            Rent Unimproved: {formatPrice(property.rent_unimproved ?? 0)}
            <br />
            Rent 1 House: {formatPrice(property.rent_one_house ?? 0)}
            <br />
            Rent 2 Houses: {formatPrice(property.rent_two_house ?? 0)}
            <br />
            Rent 3 Houses: {formatPrice(property.rent_three_house ?? 0)}
            <br />
            Rent 4 Houses: {formatPrice(property.rent_four_house ?? 0)}
            <br />
          </Text>
        </Stack>
      </Box>
      <Button
        colorScheme="green"
        isLoading={buying}
        onClick={() => {
          handleBuy();
        }}
      >
        Buy {property.name}
      </Button>
    </Box>
  );
};
