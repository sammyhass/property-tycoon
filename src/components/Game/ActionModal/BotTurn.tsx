import { TOKENS_MAP, useGameContext } from '@/hooks/useGameContext';
import { usePlayer } from '@/hooks/usePlayer';
import { formatPrice } from '@/util/formatPrice';
import {
  Alert,
  AlertDescription,
  Box,
  Button,
  chakra,
  Heading,
  ListItem,
  UnorderedList,
} from '@chakra-ui/react';
import { faCheck, faDice } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { BoardSpace, CardAction, CardType } from '@prisma/client';
import { AnimatePresence } from 'framer-motion';
import { useCallback, useEffect, useState } from 'react';

export type BotPreference =
  | 'GET_STATIONS'
  | 'GET_EXPENSIVE_SET'
  | 'GET_UTILITIES'
  | 'GET_CHEAP_SET'
  | 'MAX_MONEY_ALWAYS';

export default function BotTurn() {
  const {
    currentPlayer,
    payBank,
    rollDice,
    move,
    sendToJail,
    gameSettings,
    calculateRent,
    endTurn,
    landedOnFreeParking,
    payToFreeParking,
    totalOnFreeParking,
    couldPay,
    takeCard,
    performCardAction,
    bankrupt,
    payPlayer,
    isOwned,
    buy,
  } = useGameContext();

  // The property (if any) the bot is currently on.

  const [updates, setUpdates] = useState<(React.ReactNode | string)[]>([]);

  const [isTakingTurn, setIsTakingTurn] = useState(false);
  const [hasCompletedTurn, setHasCompletedTurn] = useState(false);
  const [hasMoved, setHasMoved] = useState(false);

  const { pos, name, getCurrentBoardSpace, getOwnedProperties } = usePlayer(
    currentPlayer?.token ?? undefined
  );

  useEffect(() => {
    setUpdates([]);
    setHasMoved(false);
    setIsTakingTurn(false);
    setHasCompletedTurn(false);
  }, [currentPlayer?.name]);

  const onLandOnGo = useCallback(() => {
    if (currentPlayer?.token) {
      payBank(currentPlayer.token, -200);
      setUpdates(updates => [
        ...updates,
        <>
          Earned <chakra.span color="green">{formatPrice(200)}</chakra.span> for
          landing on Go
        </>,
      ]);
    }
  }, [currentPlayer?.token, payBank]);

  const onLandOnProperty = useCallback(
    (propertyId: string) => {
      if (!currentPlayer?.token) return;
      const owner = isOwned(propertyId);

      const property = gameSettings?.Properties.find(
        gameProperty => gameProperty.id === propertyId
      );

      setUpdates(updates => [
        ...updates,
        <>
          Landed on <b>{property?.name}</b>
        </>,
      ]);

      // check to see if bot can purchase property
      if (!owner) {
        const canBuy = couldPay(currentPlayer?.token, property?.price ?? 0);

        if (canBuy && Math.random() > 0.3) {
          buy(currentPlayer.token, propertyId);
          setUpdates(updates => [
            ...updates,
            <>
              ðŸ’¸ Bought <b>{property?.name}</b> for{' '}
              <chakra.span color="green">
                {formatPrice(property?.price ?? 0)}
              </chakra.span>
            </>,
          ]);
        } else {
          setUpdates(updates => [
            ...updates,
            `${canBuy ? 'Chose not to buy' : 'Could not buy'} ${
              property?.name
            } for ${formatPrice(property?.price ?? 0)}`,
          ]);
        }
      } else if (owner?.token !== currentPlayer.token) {
        // Pay rent
        const rent = calculateRent(propertyId);

        if (couldPay(currentPlayer?.token, rent)) {
          payPlayer(currentPlayer.token, owner.token, rent);

          setUpdates(updates => [
            ...updates,
            <>
              Paid{' '}
              <chakra.span color="red.500">{formatPrice(rent)}</chakra.span> to
              {TOKENS_MAP[owner.token]} for landing on {property?.name}
            </>,
          ]);
        } else {
          setUpdates(updates => [
            ...updates,
            `Could not pay rent for ${property?.name}`,
            `ðŸ’€ ${name} is bankrupt!`,
          ]);
          bankrupt(currentPlayer.token);
        }
      } else {
        setUpdates(updates => [
          ...updates,
          `${name} landed on ${property?.name}, but is already the owner`,
        ]);
      }
    },
    [
      couldPay,
      gameSettings,
      currentPlayer?.token,
      isOwned,
      name,
      bankrupt,
      calculateRent,
      buy,
      payPlayer,
    ]
  );

  const onLandOnFreeParking = useCallback(() => {
    if (!currentPlayer?.token) return;
    const fpTotal = totalOnFreeParking;
    landedOnFreeParking(currentPlayer?.token);
    setUpdates(updates => [
      ...updates,
      <>
        Landed on <b>Free Parking</b> and collected{' '}
        <b>{formatPrice(fpTotal)}</b>
      </>,
    ]);
  }, [currentPlayer?.token, landedOnFreeParking]);

  const onLandOnGoToJail = useCallback(() => {
    if (!currentPlayer?.token) return;
    sendToJail(currentPlayer?.token);
  }, [currentPlayer?.token, sendToJail]);

  const onLandOnTakeCard = useCallback(
    (cardType: CardType) => {
      if (!currentPlayer?.token) return;
      let card: CardAction | null = null;
      setTimeout(() => {
        card = takeCard(cardType);
        setUpdates(updates => [
          ...updates,
          <>
            {name} took a {cardType} card - {card?.description}
          </>,
        ]);
        if (!card) return;
        performCardAction(currentPlayer?.token, card);
      }, 1000);
    },
    [currentPlayer?.token, name, takeCard, performCardAction]
  );

  const onLandOnTax = useCallback(
    (taxCost: number) => {
      if (!currentPlayer?.token) return;
      payToFreeParking(currentPlayer?.token, taxCost);
      setUpdates([
        ...updates,
        `${name} paid ${formatPrice(
          taxCost
        )} to free parking for landing on tax`,
      ]);
    },
    [payToFreeParking, currentPlayer]
  );

  const onLandedOnSpace = (space: BoardSpace) => {
    if (!currentPlayer) return;

    switch (space?.space_type) {
      case 'GO':
        onLandOnGo();
        break;
      case 'FREE_PARKING':
        onLandOnFreeParking();
        break;
      case 'GO_TO_JAIL':
        onLandOnGoToJail();
        break;
      case 'PROPERTY':
        if (!space.property_id) break;
        onLandOnProperty(space.property_id);
        break;
      case 'TAX':
        onLandOnTax(space.tax_cost ?? 0);
        break;
      case 'TAKE_CARD':
        if (!space.take_card) break;
        onLandOnTakeCard(space.take_card);
        break;
      case 'JUST_VISIT':
        setUpdates(updates => [...updates, `${name} landed on Just Visit`]);
        break;
      default:
        setHasCompletedTurn(true);
        setIsTakingTurn(false);
        break;
    }
    setTimeout(() => {
      setIsTakingTurn(false);
      setHasCompletedTurn(true);
    }, 1000);
  };

  useEffect(() => {
    if (!hasMoved) return;
    const space = getCurrentBoardSpace();
    if (space) {
      onLandedOnSpace(space);
    }
  }, [pos]);

  const performTurn = useCallback(() => {
    setIsTakingTurn(true);
    setHasCompletedTurn(false);
    if (!currentPlayer) return;

    const [dice1, dice2] = rollDice();
    setUpdates([`ðŸŽ² ${name} rolled ${dice1} and ${dice2}`]);
    if (!currentPlayer) return;
    const space = move(currentPlayer?.token, dice1 + dice2, true);
    setHasMoved(true);

    return space;
  }, [currentPlayer, rollDice, name, move, onLandedOnSpace]);

  return (
    <Box>
      <Heading size="md">
        {currentPlayer?.name}{' '}
        {!hasCompletedTurn ? 'is taking their turn...' : 'has taken their turn'}
      </Heading>
      {!isTakingTurn && !hasCompletedTurn && (
        <Alert size="sm" my="5px" borderRadius={'8px'}>
          <AlertDescription>
            Click below to simulate {name}'s turn
          </AlertDescription>
        </Alert>
      )}
      <Button
        size="lg"
        w="100%"
        my="10px"
        onClick={hasCompletedTurn ? endTurn : performTurn}
        colorScheme={hasCompletedTurn ? 'green' : 'blue'}
        isLoading={isTakingTurn}
        leftIcon={
          <FontAwesomeIcon icon={hasCompletedTurn ? faCheck : faDice} />
        }
      >
        {hasCompletedTurn ? 'End Turn' : 'Simulate Turn'}
      </Button>
      <Box mt="10px">
        <UnorderedList spacing={3} fontSize="md">
          <AnimatePresence>
            {updates.map((update, index) => (
              <ListItem
                layoutId={`${currentPlayer?.name}-update-${index}`}
                key={`${update}-${index}`}
              >
                {update}
              </ListItem>
            ))}
          </AnimatePresence>
        </UnorderedList>
      </Box>
    </Box>
  );
}
