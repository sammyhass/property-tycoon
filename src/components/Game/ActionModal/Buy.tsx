import { useGameContext } from '@/hooks/useGameContext';
import { formatPrice } from '@/util/formatPrice';
import { Box, Button, Divider, Flex, Heading } from '@chakra-ui/react';
import React, { useState } from 'react';
import { PropertyRentInfo } from '../Board/PropertyStack';
import { BoardSpaceProperty } from '../Board/spaces';

export default function BuyContent() {
  const { gameSettings, currentPlayer, state, buy, hideActionModal } =
    useGameContext();
  const [buying, setBuying] = useState(false);

  // Logic: the only way to buy a property is if it is your turn and you have landed on it.
  if (!currentPlayer) return <></>;

  const space = gameSettings?.BoardSpaces.find(
    space => space.board_position === state[currentPlayer?.token]?.pos
  );

  const property = gameSettings?.Properties.find(
    property => property.id === space?.property_id
  );

  const group = gameSettings?.PropertyGroups.find(
    group => group.color === property?.property_group_color
  );

  if (!property) return <></>;

  const handleBuy = () => {
    setBuying(true);
    buy(currentPlayer?.token, property.id);
    setTimeout(() => {
      hideActionModal();
      setBuying(false);
    }, 1000);
  };

  return (
    <Box>
      <Flex direction={'column'} justify={'center'} align="center">
        <BoardSpaceProperty property={property} />
      </Flex>
      <Divider my="15px" />
      <Box my="10px" p="10px" bg="#eee" borderRadius={'8px'}>
        <Heading size="md">
          {property.name} - {formatPrice(property.price ?? 0)}
        </Heading>
        <PropertyRentInfo property={property} group={group} />
      </Box>
      <Flex direction="column" gap="5px">
        <Button
          w="100%"
          disabled={
            !currentPlayer ||
            !property ||
            !state[currentPlayer?.token] ||
            (state[currentPlayer?.token]?.money ?? 0) < (property?.price ?? 0) // Disable the button if you don't have enough money
          }
          colorScheme="green"
          isLoading={buying}
          onClick={() => {
            handleBuy();
          }}
        >
          {(state[currentPlayer?.token]?.money ?? 0) < (property?.price ?? 0)
            ? 'Not enough money'
            : 'Buy'}
        </Button>
        <Button w="100%" colorScheme={'blue'} onClick={() => hideActionModal()}>
          No thanks
        </Button>
      </Flex>
    </Box>
  );
}
