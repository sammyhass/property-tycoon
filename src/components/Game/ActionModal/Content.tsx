import { useGameContext } from '@/hooks/useGameContext';
import { formatPrice } from '@/util/formatPrice';
import {
  Box,
  Button,
  Divider,
  Flex,
  Heading,
  Stack,
  Text,
} from '@chakra-ui/react';
import { CardAction, CardType } from '@prisma/client';
import React, { useCallback, useEffect, useState } from 'react';
import { ActionModalProps } from '.';
import GameCard from '../Board/cards/Card';
import BoardSpace from '../Board/spaces';

// Action Modal Content Components

// Content for our action modal when a player is rolling the dice.
export const ActionModalRoll = ({ onClose }: ActionModalProps) => {
  const { currentPlayer, move } = useGameContext();

  const [isRolling, setIsRolling] = useState(false);
  const [roll, setRoll] = useState(0);

  // Roll the dice.
  const handleRoll = () => {
    setIsRolling(true);

    if (!currentPlayer) return;
    // Choose a random number between 1 and 12
    const roll = Math.floor(Math.random() * 12) + 1;

    setTimeout(() => {
      move(currentPlayer?.token, roll);
      setRoll(roll);
      setIsRolling(false);
    }, 1000);

    if (!currentPlayer) {
      return <></>;
    }
  };
  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="4xl">ðŸŽ² ðŸŽ²</Heading>
      {isRolling && <Heading size="md">Rolling...</Heading>}
      {roll > 0 && <Heading size="sm">Dice landed on {roll}</Heading>}
      <Button
        w="100%"
        colorScheme="purple"
        onClick={() => handleRoll()}
        mt="10px"
      >
        Roll
      </Button>
    </Flex>
  );
};

// Content of our action modal for taking a card (pot luck or opportunity knocks)
export const ActionModalTakeCard = ({ cardType }: { cardType: CardType }) => {
  const {
    takeCard,
    payBank,
    currentPlayer,
    gameSettings,
    goto,
    hideActionModal,
  } = useGameContext();

  const [cardAction, setCardAction] = useState<CardAction | null>(null);

  const handleTakeCard = () => {
    const action = takeCard(cardType);
    if (action) {
      setCardAction(action);
    }
  };

  const performAction = useCallback(() => {
    if (!cardAction || !currentPlayer?.token) return;

    switch (cardAction.action_type) {
      case 'EARN_FROM_BANK':
        payBank(currentPlayer?.token, -(cardAction?.cost ?? 0));
        break;
      case 'PAY_BANK':
        payBank(currentPlayer?.token, cardAction?.cost ?? 0);
        break;
      case 'GO_TO_JAIL':
        const jail = gameSettings?.BoardSpaces?.find(
          space => space.space_type === 'JUST_VISIT'
        );
        if (jail) {
          goto(currentPlayer?.token, jail.board_position);
        }
        break;
      case 'GO_TO_PROPERTY':
        const property = gameSettings?.BoardSpaces?.find(
          space => space.space_type === 'PROPERTY'
        );
        if (property) {
          goto(currentPlayer?.token, property.board_position);
        }
      default:
        break;
    }

    setTimeout(() => {
      hideActionModal();
    }, 1000);
  }, [cardAction]);

  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="md">
        Pick a {cardType === 'POT_LUCK' ? 'Pot Luck' : 'Opportunity Knocks'}{' '}
        Card
      </Heading>
      {cardAction && (
        <GameCard
          {...cardAction}
          propertyName={
            gameSettings?.Properties?.find(
              property => property.id === cardAction?.property_id
            )?.name ?? undefined
          }
        />
      )}
      <Button
        colorScheme={'green'}
        mt={'10px'}
        w="100%"
        onClick={() => (cardAction ? performAction() : handleTakeCard())}
      >
        {cardAction ? 'Perform Card Action' : 'Take'}
      </Button>
    </Flex>
  );
};

export const ActionModalBuy = () => {
  const { gameSettings, currentPlayer, state, buy, hideActionModal } =
    useGameContext();
  const [buying, setBuying] = useState(false);

  // Logic: the only way to buy a property is if it is your turn and you have landed on it.

  if (!currentPlayer) return <></>;

  const space = gameSettings?.BoardSpaces.find(
    space => space.board_position === state[currentPlayer?.token]?.pos
  );

  const property = gameSettings?.Properties.find(
    property => property.id === space?.property_id
  );

  if (!property) return <></>;

  const handleBuy = () => {
    setBuying(true);
    buy(currentPlayer?.token, property.id);
    setTimeout(() => {
      hideActionModal();
      setBuying(false);
    }, 1000);
  };

  return (
    <Box>
      <Flex direction={'column'} justify={'center'} align="center">
        <BoardSpace.Property property={property} />
      </Flex>
      <Divider my="15px" />
      <Box my="10px" p="10px" bg="#eee" borderRadius={'8px'}>
        <Heading size="md">
          {property.name} - {formatPrice(property.price ?? 0)}
        </Heading>
        {property.rent_unimproved ? (
          <Stack>
            <Text p="0" fontWeight={'600'}>
              Rent Unimproved: {formatPrice(property.rent_unimproved ?? 0)}
              <br />
              Rent 1 House: {formatPrice(property.rent_one_house ?? 0)}
              <br />
              Rent 2 Houses: {formatPrice(property.rent_two_house ?? 0)}
              <br />
              Rent 3 Houses: {formatPrice(property.rent_three_house ?? 0)}
              <br />
              Rent 4 Houses: {formatPrice(property.rent_four_house ?? 0)}
              <br />
            </Text>
          </Stack>
        ) : null}
      </Box>
      <Flex direction="column" gap="5px">
        <Button
          w="100%"
          disabled={
            !currentPlayer ||
            !property ||
            !state[currentPlayer?.token] ||
            (state[currentPlayer?.token]?.money ?? 0) < (property?.price ?? 0) // Disable the button if you don't have enough money
          }
          colorScheme="green"
          isLoading={buying}
          onClick={() => {
            handleBuy();
          }}
        >
          {(state[currentPlayer?.token]?.money ?? 0) < (property?.price ?? 0)
            ? 'Not enough money'
            : 'Buy'}
        </Button>
        <Button w="100%" colorScheme={'blue'} onClick={() => hideActionModal()}>
          No thanks
        </Button>
      </Flex>
    </Box>
  );
};

export const ActionModalTax = () => {
  const { gameSettings, state, currentPlayer, payBank, hideActionModal } =
    useGameContext();

  if (!gameSettings || !currentPlayer) return <></>;

  const space = gameSettings?.BoardSpaces.find(
    space => space.board_position === (state[currentPlayer?.token]?.pos ?? 0)
  );

  const taxAmount = space?.tax_cost ?? 0;

  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="md">Pay the Tax</Heading>
      <Text fontSize={'4xl'}>{formatPrice(taxAmount)}</Text>
      <Button
        w={'100%'}
        colorScheme={'red'}
        onClick={() => {
          payBank(currentPlayer?.token, taxAmount);
          setTimeout(() => {
            hideActionModal();
          }, 1000);
        }}
      >
        Pay
      </Button>
    </Flex>
  );
};

export const ActionModalPayRent = () => {
  const {
    payPlayer,
    gameSettings,
    currentPlayer,
    state,
    isOwned,
    hideActionModal,
  } = useGameContext();

  if (!gameSettings || !currentPlayer) return <></>;

  const space = gameSettings?.BoardSpaces.find(
    space => space.board_position === (state[currentPlayer?.token]?.pos ?? 0)
  );

  const property = gameSettings?.Properties.find(
    property => property.id === space?.property_id
  );

  if (!property) return <></>;

  const owner = isOwned(property?.id);

  if (!owner) return <></>;

  const rent = property?.rent_unimproved ?? 0;

  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="md">Pay the Rent</Heading>
      {property && <BoardSpace.Property property={property} />}
      <Text fontSize={'4xl'}>{formatPrice(rent ?? 0)}</Text>
      <Button
        w={'100%'}
        colorScheme={'red'}
        onClick={() => {
          payPlayer(currentPlayer?.token, owner?.token, rent);

          setTimeout(() => {
            hideActionModal();
          }, 1000);
        }}
      >
        Pay
      </Button>
    </Flex>
  );
};

export const ActionModalFreeParking = () => {
  const {
    landedOnFreeParking,
    currentPlayer,
    hideActionModal,
    totalOnFreeParking,
  } = useGameContext();

  useEffect(() => {
    if (!currentPlayer) return;
    const v = landedOnFreeParking(currentPlayer.token);
  }, [currentPlayer]);

  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="md">Free Parking</Heading>

      <Text>
        The total amount on free parking is {formatPrice(totalOnFreeParking)},
        It&apos;s all yours now
      </Text>
      <Button
        w={'100%'}
        colorScheme={'blue'}
        onClick={() => {
          hideActionModal();
        }}
      >
        Continue
      </Button>
    </Flex>
  );
};

export const ActionModalGoToJail = () => {
  const { sendToJail, currentPlayer, hideActionModal, gameSettings } =
    useGameContext();

  const handleGoToJail = () => {
    if (!currentPlayer) return;
    sendToJail(currentPlayer?.token);

    setTimeout(() => {
      hideActionModal();
    }, 1000);
  };

  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="md">Go To Jail</Heading>
      <Text>You have been sent to jail</Text>
      <Button onClick={() => handleGoToJail()}>Go To Jail</Button>
    </Flex>
  );
};
