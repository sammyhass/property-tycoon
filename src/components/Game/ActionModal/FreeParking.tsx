import { useGameContext } from '@/hooks/useGameContext';
import { formatPrice } from '@/util/formatPrice';
import { Button, Flex, Heading, Text } from '@chakra-ui/react';
import React, { useEffect } from 'react';

export default function ActionModalFreeParking() {
  const {
    landedOnFreeParking,
    currentPlayer,
    hideActionModal,
    totalOnFreeParking,
  } = useGameContext();

  useEffect(() => {
    if (!currentPlayer) return;
    const v = landedOnFreeParking(currentPlayer.token);
  }, [currentPlayer, landedOnFreeParking]);

  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="md">Free Parking</Heading>

      <Text>
        The total amount on free parking is {formatPrice(totalOnFreeParking)},
        It&apos;s all yours now
      </Text>
      <Button
        w={'100%'}
        colorScheme={'blue'}
        onClick={() => {
          hideActionModal();
        }}
      >
        Continue
      </Button>
    </Flex>
  );
}
