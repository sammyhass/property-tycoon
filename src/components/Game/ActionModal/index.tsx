import { Box, Heading } from '@chakra-ui/react';
import React, { useMemo } from 'react';
import HUD from '../HUD';
import {
  ActionModalBuy,
  ActionModalFreeParking,
  ActionModalGetOutJail,
  ActionModalGo,
  ActionModalGoToJail,
  ActionModalPayRent,
  ActionModalRoll,
  ActionModalTakeCard,
  ActionModalTax,
} from './Content';

// Different types of action modal we can use for our modal.
export type ActionType =
  | 'BUY' // Buying a property
  | 'PAY_RENT' // Pay rent
  | 'ROLL' // Roll dice
  // Taking Cards
  | 'TAKE_POT_LUCK'
  | 'TAKE_OPPORTUNITY_KNOCKS'
  // ---
  | 'GO_TO_JAIL' // Player is sent to jail
  | 'GET_OUT_OF_JAIL' // Player attempts to get out of jail (shows instead)
  | 'FREE_PARKING' // Land on a Free Parking space
  | 'TAX' // Pay tax
  // TODO:
  | 'MORTGAGE'
  | 'GO'; // Land on a Go space

export interface ActionModalProps {
  action: ActionType | null;
}

const actionModalComponents: Record<ActionType, React.FC<ActionModalProps>> = {
  BUY: ActionModalBuy,
  MORTGAGE: () => <div>Mortgage</div>,
  PAY_RENT: ActionModalPayRent,
  GET_OUT_OF_JAIL: ActionModalGetOutJail,
  ROLL: ActionModalRoll,
  GO_TO_JAIL: ActionModalGoToJail,
  TAKE_POT_LUCK: () => <ActionModalTakeCard cardType={'POT_LUCK'} />,
  TAKE_OPPORTUNITY_KNOCKS: () => (
    <ActionModalTakeCard cardType="OPPORTUNITY_KNOCKS" />
  ),
  FREE_PARKING: ActionModalFreeParking,
  GO: ActionModalGo,
  TAX: ActionModalTax,
};

export default function ActionSidebar(props: ActionModalProps) {
  const ActionComponent = useMemo(
    () =>
      props.action
        ? (actionModalComponents[props.action] as React.FC<ActionModalProps>)
        : null,
    [props.action]
  );

  return (
    <Box minW={'400px'} bg="white" px="20px" h="100vh">
      <Heading>
        {props.action
          ?.split('_')
          .map(w => w[0].toUpperCase() + w.slice(1).toLowerCase())
          .join(' ')}
      </Heading>
      <Box minH="70%">{ActionComponent && <ActionComponent {...props} />}</Box>
      <HUD />
    </Box>
  );
}
