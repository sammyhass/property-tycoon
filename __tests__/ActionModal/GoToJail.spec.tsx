import { ActionModalGoToJail } from '@/components/Game/ActionModal/Content';
import { PlayerState } from '@/hooks/useGameContext';
import { act } from '@testing-library/react';
import React from 'react';
import { fakePlayer, renderWithGameContext } from '../fakers';

describe('Go to Jail ActionModal', () => {
  it('matches snapshot', () => {
    const { container } = renderWithGameContext(<ActionModalGoToJail />, {
      currentPlayer: fakePlayer('cat'),
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the correct text', () => {
    const { container } = renderWithGameContext(<ActionModalGoToJail />, {
      currentPlayer: fakePlayer('boot'),
    });

    expect(container.textContent?.toUpperCase()).toContain('GO TO JAIL');
    expect(container.textContent?.toUpperCase()).toContain(
      'YOU HAVE BEEN SENT TO JAIL'
    );
  });

  it('after clicking button, player is sent to jail', async () => {
    let state: PlayerState = {
      cat: {
        inJail: false,
        money: 1200,
        pos: 3,
        propertiesOwned: [],
      },
      boot: {
        inJail: false,
        money: 400,
        pos: 1,
        propertiesOwned: [],
      },
    };

    const sendToJail = jest.fn();

    const { container } = renderWithGameContext(<ActionModalGoToJail />, {
      currentPlayer: fakePlayer('cat'),
      state,
      sendToJail,
    });

    const btn = container.querySelector('button');

    if (!btn) throw new Error('No button found');

    await act(async () => {
      btn.click();

      await new Promise(resolve => setTimeout(resolve, 1200));
    });

    expect(sendToJail).toHaveBeenCalledWith('cat');
  });
});
