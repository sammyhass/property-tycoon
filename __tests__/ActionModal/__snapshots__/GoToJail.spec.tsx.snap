// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Go to Jail ActionModal matches snapshot 1`] = `
<div
  class="css-u6v8er"
>
  <h2
    class="chakra-heading css-0"
  >
    Go To Jail
  </h2>
  <p
    class="chakra-text css-0"
  >
    You have been sent to jail
  </p>
  <button
    class="chakra-button css-5c2176"
    type="button"
  >
    Go To Jail
  </button>
</div>
`;
